lib/VolSurface/Calibration/Equities.pm

```diff
@@ -0,0 +1,637 @@
+);
+
+has _scale => (
+    is      => 'ro',
+    isa     => 'Num',
+    default => 10,
+);
+
+=head2 function_to_optimize
+
+The function that we want to optimize.
+
+=cut
+
+sub function_to_optimize {
+    my ($self, $params) = @_;
+
+    # this number is returned when the optimization is heading the wrong path
+    my $error = 1000;
+
+    my %params =
+        pairwise { our $a => our $b } @{$self->calibration_param_names},
+        @$params;
+
+    my @tenors = @{$self->term_by_day};
+    my @kurtosis =
+        map { $self->_calculate_kurtosis($_ / 365, \%params) } @tenors;
+    my @variance =
+        map { $self->_calculate_variance($_ / 365, \%params) } @tenors;
+    my @skew = map { $self->_calculate_skew($_ / 365, \%params) } @tenors;
+
+    foreach my $param_name (keys %params) {
+        my $param_value = $params{$param_name};
+        return $error
+            if ($param_name =~ /wing(R|L)/i && max(abs($param_value)) > 20.0);
+        return $error
+            if ($param_name =~ /growth/i && max(abs($param_value)) > 5.0);
+        return $error
+            if ($param_name !~ /wing(R|L)/i
+            && $param_name !~ /growth/i
+            && max(abs($param_value)) > 2.0);
+    }
+
+    return $error if (min(@variance) < 0);
+    my @atmvols            = map { sqrt($_) } @variance;
+    my $calibrated_surface = $self->get_calibrated_surface(\%params);
+    my $actual_surface     = $self->surface;
+
+    my $total = 0;
+    foreach my $day (@{$self->term_by_day}) {
+        my $sum       = 0;
+        my $tenorvega = 0;
+        my $atm_vol   = shift @atmvols;
+        foreach my $point (@{$self->smile_points}) {
+            my $atm_check      = log($point / 100) / $atm_vol / sqrt($day / 365);
+            my $calibrated_vol = $calibrated_surface->{$day}->{$point};
+            my $actual_vol     = $actual_surface->{$day}->{smile}->{$point};
+            my $d1             = (log(100 / $point) + (0.5 * $actual_vol**2) * ($day / 365)) / ($actual_vol * ($day / 365)**0.5);
+            my $nd1            = (1 / (2 * 3.1416)**0.5) * exp(-$d1 * $d1 / 2);
+            my $vega           = $nd1 * ($day / 365)**0.5;
+            $sum += $vega * abs($calibrated_vol - $actual_vol);
+            $tenorvega += $vega;
+        }
+        $total += $sum / $tenorvega;
+    }
+
+    return $total;
+}
+
+=head2 calibration_param_names
+
+Calibration parameter names.
+It is hard-coded here because it needs to be in this sequence
+
+=cut
+
+has calibration_param_names => (
+    is      => 'ro',
+    isa     => 'ArrayRef',
+    default => sub {
+        [qw(
+                atmvolshort
+                atmvol1year
+                atmvolLong
+                atmWingL
+                atmWingR
+                skewshort
+                skew1year
+                skewlong
+                skewwingL
+                skewwingR
+                kurtosisshort
+                kurtosislong
+                kurtosisgrowth
+                )];
+    },
+);
```

Comment:
The module quite often relies on that order. Yet, you always work with hashes. I think the module can benefit both in performance and readability from changing that pattern to use a list instead of a hash.

```perl
use constant {
    ATMVOLSHORT => 0,
    ATMVOL1YEAR => 1,
    ...
    KURTOSISGROWTH => 12,
};

sub function_to_optimize {
    my ($self, $params) = @_;

    # this number is returned when the optimization is heading the wrong path
    my $error = 1000;

    for my $i (ATMWINGL, ATMWINGR, SKEWWINGL, SKEWWINGR) {
        return $error if abs($params->[$i]) > 20;
    }

    return $error if abs($params->[KURTOSISGROWTH]) > 5;

    for my $i (ATMSHORT, ATM1YEAR, ATMLONG,
               SKEWSHORT, SKEW1YEAR, SKEWLONG,
               KURTOSISSHORT, KURTOSISLONG) {
        return $error if abs($params->[$i]) > 2;
    }

    my @tenors = @{$self->term_by_day};

    my @variance = $self->_calculate_variance(@tenors, $params);
    return $error if (min(@variance) < 0);

    my @kurtosis = $self->_calculate_kurtosis(@tenors, $params);
    my @skew = $self->_calculate_skew(@tenors, $params);

    my @atmvols            = map { sqrt($_) } @variance;
    my $calibrated_surface = $self->get_calibrated_surface($params);
    my $actual_surface     = $self->surface;

    my $total = 0;
...
    return $total;
}

sub _get_params_from {
    my ($self, $params) = @_;

    return [map { roundnear(0.0001, $_) } @$params];
}

sub _calculate_kurtosis {
    my $self = shift;
    my $params = pop;

    my $kurt_small = $params->[KURTOSISSHORT];
    my $kurt_alpha = $params->[KURTOSISGROWTH];
    my $kurt_inf   = $params->[KURTOSISLONG];

    return map {$kurt_small + ($kurt_inf - $kurt_small) * (1 - exp(-$kurt_alpha * $_ / 365))} @_;
}
```